### PR TITLE
[Android] Split code into two stores

### PR DESCRIPTION
**Pull request by andresesfm**
Work in Progress.

Creates two variants: `play` and `amazon` and only uses the required code.

NOTE: This would be a breaking change with a very simple fix described in the documentation. To add: `missingDimensionStrategy 'store', 'play'` `in build.gradle`

### CODE REVIEW

File: src/iap.ts [hunk 1]
```diff
@@ -204,7 +204,7 @@ export const flushFailedPurchasesCachedAsPendingAndroid = (): Promise<
 
       await checkNativeAndroidAvailable(myRNIapModule);
 
-      return RNIapModule.flushFailedPurchasesCachedAsPending();
+      return RNIapModule ? RNIapModule.flushFailedPurchasesCachedAsPending() :[];
```
Comment (andresesfm) [inline src/iap.ts:207]: can be improved later

File: .gitignore [hunk 2]
```diff
@@ -55,6 +55,13 @@ react-native-iap-*.tgz
 # Built files
 tsconfig.tsbuildinfo
 
+# Generated files
 *.js.flow
 *.d.ts
-*.js
```
Comment (andresesfm) [inline .gitignore:60]: This makes it difficult to search inside IapExample folder since vscode ignores all the js.

File: IapExample/App.js [hunk 3]
```diff
@@ -115,27 +115,21 @@ class Page extends Component {
 
   async componentDidMount(): void {
     try {
-      const result = await RNIap.initConnection();
```
Comment (andresesfm) [inline IapExample/App.js:118]: returns void, no need to log it

File: IapExample/App.js [hunk 4]
```diff
@@ -115,27 +115,21 @@ class Page extends Component {
...
-      console.log('result', result);
     } catch (err) {
       console.warn(err.code, err.message);
     }
 
     purchaseUpdateSubscription = purchaseUpdatedListener(
       async (purchase: InAppPurchase | SubscriptionPurchase) => {
-        const receipt = purchase.transactionReceipt;
+        console.info("purchase",purchase);
+        const receipt =  purchase.transactionReceipt? purchase.transactionReceipt : purchase.originalJson;
```
Comment (andresesfm) [inline IapExample/App.js:127]: receipt is named something different on Amazon and Google

File: IapExample/android/app/build.gradle [hunk 5]
```diff
@@ -162,6 +162,17 @@ android {
             proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
         }
     }
+    flavorDimensions "appstore"
```
Comment (andresesfm) [inline IapExample/android/app/build.gradle:165]: An example where the dimension name in the client app is different from the one in the library

File: IapExample/android/app/build.gradle [hunk 6]
```diff
@@ -180,7 +191,8 @@ android {
 
 dependencies {
     implementation fileTree(dir: "libs", include: ["*.jar"])
-    implementation "com.facebook.react:react-native:+"  // From node_modules
+    implementation "com.facebook.react:react-native:+"
+    implementation project(path: ':react-native-iap')
```
Comment (andresesfm) [inline IapExample/android/app/build.gradle:195]: pointing to local file for quick debug

File: IapExample/android/settings.gradle [hunk 7]
```diff
@@ -1,3 +1,5 @@
 rootProject.name = 'IapExample'
 apply from: file("../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesSettingsGradle(settings)
 include ':app'
+include ':react-native-iap'
```
Comment (andresesfm) [inline IapExample/android/settings.gradle:4]: pointing to local for quick debug iteration

File: IapExample/package.json [hunk 8]
```diff
@@ -13,7 +13,7 @@
     "apsl-react-native-button": "^3.1.1",
     "react": "16.9.0",
     "react-native": "0.61.5",
-    "react-native-iap": "^6.0.6",
+    "react-native-iap": "file:..",
```
Comment (andresesfm) [inline IapExample/package.json:16]: pointing to local for quick debug iteration

File: README.md [hunk 9]
```diff
@@ -246,7 +260,7 @@ Linking the package manually is not required anymore with [Autolinking](https://
     }
     -keep enum org.greenrobot.eventbus.ThreadMode { *; }
     ```
-5. Add the following to the `<permission>` block in `android/app/src/main/AndroidManifest.xml`:
```
Comment (andresesfm) [inline README.md:249]: No longer needed as it's part of the new implementation

File: README_AMAZON.md [hunk 10]
```diff
@@ -15,19 +15,6 @@ Here are the additional steps to add Amazon IAP support.
   -keepattributes *Annotation*
   ```
 
-  2. Update AndroidManifest.xml to add the following:
```
Comment (andresesfm) [inline README_AMAZON.md:18]: No longer needed, it's part of the library implementation now

File: android/build.gradle [hunk 11]
```diff
@@ -46,8 +55,8 @@ repositories {
 
 dependencies {
   implementation 'com.facebook.react:react-native:+'
-  implementation 'com.android.billingclient:billing:3.0.3'
-  implementation fileTree(dir: 'libs', include: ['*.jar'])
+  playImplementation 'com.android.billingclient:billing:3.0.3'
+  amazonImplementation fileTree(dir: 'libs', include: ['*.jar'])
```
Comment (andresesfm) [inline android/build.gradle:59]: Splitting libraries per flavor. Reduces the bundle size and avoids bugs like the one we are seeing with OneSignal

File: android/src/amazon/java/com/dooboolab/RNIap/RNIapAmazonModule.java [hunk 12]
```diff
@@ -53,41 +53,31 @@
...
 
-  private LifecycleEventListener lifecycleEventListener;
-  private PurchasingListener purchasingListener = null;
 
   public RNIapAmazonModule(final ReactApplicationContext reactContext) {
     super(reactContext);
-    this.reactContext = reactContext;
 
-    lifecycleEventListener =
+    UiThreadUtil.runOnUiThread(() -> {
```
Comment (andresesfm) [inline android/src/amazon/java/com/dooboolab/RNIap/RNIapAmazonModule.java:61]: This new implementation follows Amazon's documentation guidelines
https://developer.amazon.com/docs/in-app-purchasing/iap-implement-iap.html#4-implement-getpurchaseupdates-method

File: android/src/amazon/java/com/dooboolab/RNIap/RNIapAmazonModule.java [hunk 13]
```diff
@@ -135,10 +125,11 @@ public void getItemsByType(
     for (int ii = 0, skuSize = skuArr.size(); ii < skuSize; ii++) {
       productSkus.add(skuArr.getString(ii));
     }
-    RequestId requestId = PurchasingService.getProductData(productSkus);
     DoobooUtils
-      .getInstance()
-      .addPromiseForKey(PROMISE_GET_PRODUCT_DATA, promise);
+            .getInstance()
+            .addPromiseForKey(PROMISE_GET_PRODUCT_DATA, promise);
+    RequestId requestId = PurchasingService.getProductData(productSkus);
```
Comment (andresesfm) [inline android/src/amazon/java/com/dooboolab/RNIap/RNIapAmazonModule.java:131]: Change in order prevents unlikely race condition

File: src/iap.ts [hunk 14]
```diff
@@ -48,7 +48,7 @@ export function getInstallSourceAndroid(): InstallSourceAndroid {
 }
 
 async function detectInstallSourceAndroid(): Promise<void> {
```
Comment (andresesfm) [inline src/iap.ts:50]: The method could be reimplemented in a simplified way, TODO

File: src/iap.ts [hunk 15]
```diff
@@ -805,13 +810,13 @@ export const purchaseUpdatedListener = (
 
     return myModuleEvt.addListener('purchase-updated', listener);
   } else {
-    const emitterSubscription = DeviceEventEmitter.addListener(
```
Comment (andresesfm) [inline src/iap.ts:808]: It is now deprecated

File: android/build.gradle (hunk 11 above)
Comment (andresesfm) [reply android/build.gradle:59]: This crash is resolved with this PR: https://github.com/dooboolab/react-native-iap/issues/1291

**Review by hyochan — CHANGES_REQUESTED** [review]
Thanks for real hard work @andresesfm. You've rocked the `Amazon` iap.

I've just given small reviews and would like to test my previous android app if things aren't breaking after changes.

Also, since this is pretty much a heavy update, I would prefer you to update the version to `6.1.0` instead of `6.0.10`.

**Review by hyochan — COMMENTED** [review]
@andresesfm Hi. I think you're done with the review fixes and I've just tried installing this new version in [dooboolab.com pull request](https://github.com/hyochan/dooboolab.com/pull/16) by `yarn add react-native-iap@next` which is currently version `6.1.0`. However, I am facing the below build error. Could you kindly look into this and see what should be done to solve the build problem?

I am afraid if we are missing additional context for the previous user. The above project was using `6.0.3` which did not have a build problem.

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':app:mergeDebugAssets'.
> Could not resolve all task dependencies for configuration ':app:debugRuntimeClasspath'.
   > Could not resolve project :react-native-iap.
     Required by:
         project :app
      > The consumer was configured to find a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'. However we cannot choose between the following variants of project :react-native-iap:
          - amazonDebugRuntimeElements
          - playDebugRuntimeElements
        All of them match the consumer attributes:
          - Variant 'amazonDebugRuntimeElements' capability dooboolab:react-native-iap:unspecified declares a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Unmatched attributes:
                  - Provides attribute 'com.android.build.api.attributes.VariantAttr' with value 'amazonDebug' but the consumer didn't ask for it
                  - Provides attribute 'store' with value 'amazon' but the consumer didn't ask for it
          - Variant 'playDebugRuntimeElements' capability dooboolab:react-native-iap:unspecified declares a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Unmatched attributes:
                  - Provides attribute 'com.android.build.api.attributes.VariantAttr' with value 'playDebug' but the consumer didn't ask for it
                  - Provides attribute 'store' with value 'play' but the consumer didn't ask for it
        The following variants were also considered but didn't match the requested attributes:
          - Variant 'amazonDebugApiElements' capability dooboolab:react-native-iap:unspecified declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Incompatible because this component declares an API of a component and the consumer needed a runtime of a component
          - Variant 'amazonReleaseApiElements' capability dooboolab:react-native-iap:unspecified:
              - Incompatible because this component declares an API of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
          - Variant 'amazonReleaseRuntimeElements' capability dooboolab:react-native-iap:unspecified declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
          - Variant 'playDebugApiElements' capability dooboolab:react-native-iap:unspecified declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Incompatible because this component declares an API of a component and the consumer needed a runtime of a component
          - Variant 'playReleaseApiElements' capability dooboolab:react-native-iap:unspecified:
              - Incompatible because this component declares an API of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
          - Variant 'playReleaseRuntimeElements' capability dooboolab:react-native-iap:unspecified declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 24s

error Failed to install the app. Make sure you have the Android development environment set up: https://reactnative.dev/docs/environment-setup.
Error: Command failed: ./gradlew app:installDebug -PreactNativeDevServerPort=8081

FAILURE: Build failed with an exception.
```

**Review by hyochan — APPROVED** [review]
I've just find out the solution in the `README.md`.

```
missingDimensionStrategy 'store', 'play'
```

Thanks a lot for your contribution :pray: 
I also like to reward you for this `PR`. 

Could you submit expense in [react-native-iap opencollective](https://opencollective.com/react-native-iap)? I want to reward you with $50. know it isn't much but I want to thank you for helping react-native-iap.

**Pull request merged**